### ISSUE TITLE
Redirect to previous page when login on the website

### ISSUE THREAD

**Issue by vlegoff** [evennia/evennia#1576]
#### Brief summary of issue / Description of requested feature:

When one logs in the website with his account name and password, he is redirected to the home page.  It would make more sense to redirect him to the page on which he was before pressing the login link, as he might be prevented from accessing this page without login in.

#### Steps to reproduce the issue / Reasons for adding feature:

1. It is probably quite easy to do.
2. I don't suppose it would annoy other game developers/current players.
3. If the redirect page can be set in the settings (I believe so), perhaps a special symbol, like `:back` (that can't be a proper file name) could be used to indicate redirecting the user to the previous page.

#### Extra information, such as Evennia revision/repo/branch, operating system and ideas for how to solve / implement:

* Evennia: develop.

**Comment by chainsol** [comment]
Hmm, this would just involve either passing the current page to the login view or storing it in the session. Should be relatively easy!

**Comment by vlegoff** [comment]
I guess it would.  But I don't know enough Django to do it with confidence.  Do you think it could (should) be implemented in Evennia vanilla, or could it (should it) be added in individual games?

**Comment by Griatch** [comment]
We are using the default Django login functionality here, haven't looked into how that would be done.

**Comment by chainsol** [comment]
Ah, turns out we only need to pass an argument to the `login_required` decorator, `login_url`, to tell Django to redirect somewhere after login. In `settings.py`, one can actually set this as a default for all `login_required` decorators, or one can set it per decorator as per [the Django docs.](https://docs.djangoproject.com/en/2.0/topics/auth/default/#the-login-required-decorator)

We use this type of decorator only once - in the link to the admin page, we have `staff_member_required`, which is actually the same as `login_required` as far as how to redirect correctly.

Basically - if you use `login_required` or `staff_member_required`, use it like:
```
login_required(login_url = "/wherever/you/should/go/after/login")
def your_view_here(request, *args, **kwargs):
```

**Comment by TehomCD** [comment]
It looks like the `@login_required` decorator has its default behavior as going back to the page they were attempting to go to before being prompted for a login. Specifying the login_url keyword argument is to change the url of the view you want to use for the login page, it looks like. So just attaching `@login_required` before your view should do what you want, I think. Maybe a view is just manually redirecting to the login page without using the decorator?

**Comment by vlegoff** [comment]
Ha, thanks!  So it's not an Evennia issue and could be closed right away.  I would have liked this behavior for the wiki pages, but as login may not be necessarily needed, perhaps using `@login_required` wouldn't do.  I was thinking about remembering the page from which the login link is clicked and going back to it afterwards, but I realize it doesn't make that much sense.  I will think about what should be done for the Evennia wiki.  Perhaps `login_required` can be used as a conditional function rather than a systematic decorator.  Thanks again!

**Comment by TehomCD** [comment]
What you might be able to do is pass it along with the URL for the login page from whatever template you wanna redirect back to it. It sounds like you could do ` "/accounts/login/?next={{current_page_url}}"`. I haven't tried it myself, but it looks like anything passed as the 'next' argument in its GET request will do a redirect to that URL. So you could probably add something like `current_page_url = request.get_full_path()` to your template and then append that to the login url as the value of 'next', maybe.

**Comment by chainsol** [comment]
I believe that we changed the behavior you describe, @TehomCD, since in our [`settings_default.py`](https://github.com/evennia/evennia/blob/master/evennia/settings_default.py#L682) changes the default redirect to be `"/"`. Perhaps in stock Django that'd be the default, though, but it looks like we default to redirecting to home.

**Comment by TehomCD** [comment]
[Django's docs](https://docs.djangoproject.com/en/2.0/ref/settings/#login-redirect-url) states that the setting is used when the view gets no `next` parameter, so it's ignored when you specify that.

**Comment by chainsol** [comment]
Right, very true, I just meant that the login_required decorator itself will always redirect to home in Evennia, unless you provide your own url, either via next in the link itself or passing login_url in the decorator.

**Comment by TehomCD** [comment]
That doesn't appear to be correct when I used it. `@login_required` redirects just fine to the page I was attempting to log into without passing an argument to it, and I haven't changed the setting. I believe that it gets `next` as a default when using the decorator - I don't know how you can disable the `next` behavior when using the decorator at all.

**Comment by chainsol** [comment]
Ooh, that's interesting. You used a link with next as a parameter? Or you're just using login_required and it still redirects correctly?

**Comment by TehomCD** [comment]
Just using `@login_required` for function-based views and it automatically redirects correctly. The docs say:
```
If the user isn’t logged in, redirect to settings.LOGIN_URL, passing the current absolute path in the query string. Example: /accounts/login/?next=/polls/3/.
```
And then the template has it pass along {{ next }} in the form. I'm guessing if you wanted to prevent the next behavior you'd probably have to snip that out of the template, but I don't know if there's a way to suppress the redirection on a per-use basis of the decorator that doesn't cause you to get a 400 error. Either way, it seems like it does what people usually want it to when just used in its default case.

For the `base.html` template I did the example above where I passed along `?next={{ request.get_absolute_path|urlencode }}` to the login link which we can do since evennia added `'django.contrib.auth.context_processors.auth',`, which makes `{{ request }}` accessible. But the decorator doesn't require that sort of thing, it just works automatically.

**Comment by vlegoff** [comment]
Thanks!  I guess in my case what I should do is simply to redirect to the `settings.LOGIN_URL` with next being the same URL if the user isn't authenticated, and the wiki page she's trying to read isn't public.  It might solve my problem.  I'll check!
